### PR TITLE
Deploy mongo as pacemaker resource

### PR DESCRIPTION
**Pull request by patelb10**

### CODE REVIEW

**Comment by johren** [comment]
Can you provide the output of running this test, please?

**Comment by patelb10** [comment]
python run_tests.py -stack 14 -config config-mn -test deploy/rackhd_ha_resource_install.py -numvms 4
*** Using config file path: config-mn
*** Created config file: config-mn/generated/fit-config-20170310-213415
*** Using config file: config-mn/generated/fit-config-20170310-213415
2017-03-10 21:34:17,625 INFO    removing previous logging dir /emc/patelb10/RAC-4243/RackHD/test/log_output/run_2017-03-09_15:39:18.d 3711 MainProcess stream-monitor/flogging/infra_logging.py:__init__@131 gl-main
2017-03-10 21:34:17,625 INFO    this runs logging dir /emc/patelb10/RAC-4243/RackHD/test/log_output/run_2017-03-10_21:34:17.d 3711 MainProcess stream-monitor/flogging/infra_logging.py:__init__@131 gl-main
2017-03-10 21:34:17,626 INFO     Start Of Test Block: 1
*** Reloading config file: config-mn/generated/fit-config-20170310-213415
2017-03-10 21:34:18,246 INFO    +1.01 - STARTING TEST: test01_install_mongo_resource (test.deploy.rackhd_ha_resource_install.rackhd_ha_resource_install)
test01_install_mongo_resource (test.deploy.rackhd_ha_resource_install.rackhd_ha_resource_install) ... VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm_mon -X | grep 'node1.*online=.true' -q
crm_mon -X | grep 'node1.*online=.true' -q
remote_shell: Exit Code = 0
VM number:  2
remote_shell: Host = stack14-ora-1.admin
remote_shell: Command = crm_mon -X | grep 'node2.*online=.true' -q
crm_mon -X | grep 'node2.*online=.true' -q
remote_shell: Exit Code = 0
VM number:  3
remote_shell: Host = stack14-ora-2.admin
remote_shell: Command = crm_mon -X | grep 'node3.*online=.true' -q
crm_mon -X | grep 'node3.*online=.true' -q
remote_shell: Exit Code = 0
VM number:  4
remote_shell: Host = stack14-ora-3.admin
remote_shell: Command = crm_mon -X | grep 'node4.*online=.true' -q
crm_mon -X | grep 'node4.*online=.true' -q
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive docker_mongo_1 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
crm configure primitive docker_mongo_1 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive mongo_addr_1 ocf:heartbeat:IPaddr2 params ip='172.31.128.121' nic='ens192' op monitor interval='10s' meta is-managed='true'
crm configure primitive mongo_addr_1 ocf:heartbeat:IPaddr2 params ip='172.31.128.121' nic='ens192' op monitor interval='10s' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure colocation mongo1 inf: docker_mongo_1 mongo_addr_1
crm configure colocation mongo1 inf: docker_mongo_1 mongo_addr_1
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive docker_mongo_2 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
crm configure primitive docker_mongo_2 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive mongo_addr_2 ocf:heartbeat:IPaddr2 params ip='172.31.128.122' nic='ens192' op monitor interval='10s' meta is-managed='true'
crm configure primitive mongo_addr_2 ocf:heartbeat:IPaddr2 params ip='172.31.128.122' nic='ens192' op monitor interval='10s' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure colocation mongo2 inf: docker_mongo_2 mongo_addr_2
crm configure colocation mongo2 inf: docker_mongo_2 mongo_addr_2
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive docker_mongo_3 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
crm configure primitive docker_mongo_3 ocf:heartbeat:docker params allow_pull=true image='registry.hwimo.lab.emc.com/mongo' run_opts=\'--privileged=true --net='host' -d -p 27017:27017\' run_cmd=\'--replSet mongo_rs --logpath /var/log/mongodb/mongod.log\' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure primitive mongo_addr_3 ocf:heartbeat:IPaddr2 params ip='172.31.128.123' nic='ens192' op monitor interval='10s' meta is-managed='true'
crm configure primitive mongo_addr_3 ocf:heartbeat:IPaddr2 params ip='172.31.128.123' nic='ens192' op monitor interval='10s' meta is-managed='true'
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = crm configure colocation mongo3 inf: docker_mongo_3 mongo_addr_3
crm configure colocation mongo3 inf: docker_mongo_3 mongo_addr_3
remote_shell: Exit Code = 0
scp_file_to_host: 'scp -o StrictHostKeyChecking=no mongo_replica_init.py onrack@stack14-ora.admin:'
scp_file_to_host: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = chmod 777 mongo_replica_init.py
chmod 777 mongo_replica_init.py
remote_shell: Exit Code = 0
VM number:  1
remote_shell: Host = stack14-ora.admin
remote_shell: Command = ./mongo_replica_init.py
./mongo_replica_init.py
remote_shell: Exit Code = 0
ok
2017-03-10 21:34:43,984 INFO    -1.01 - ENDING TEST: test01_install_mongo_resource (test.deploy.rackhd_ha_resource_install.rackhd_ha_resource_install)

----------------------------------------------------------------------
Ran 1 test in 25.744s

OK


@johren

File: test/config_templates/mongo_init.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+import pymongo
+
+MONGODB_URI = 'mongodb://172.31.128.121:27017/admin'
+
+
+def main():
+
+    client = pymongo.MongoClient(MONGODB_URI)
+
+    config = {{mongo_list}}
```
Comment (stuart-stanley) [inline test/config_templates/mongo_init.py]: You can silence hound on a case-by-case basis by adding a comment at the end of the line like:
```    config = {{mongo_list}}  # NOQA: undefined name```

File: test/deploy/rackhd_ha_resource_install.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+import fit_common
+
+nodelist = []   # list of active nodes in cluster
+numvms = int(fit_common.fitargs()['numvms'])
+err = []
+numrs = numvms - 1   # number of mongo resource
+vip_dict = {'mongo': [], 'rabbit': [], 'rackhd': []}
+
+
+class rackhd_ha_resource_install(unittest.TestCase):
```
Comment (stuart-stanley) [inline test/deploy/rackhd_ha_resource_install.py]: please add the line "   longMessage = True" here....

Doing so means that self.AssertXXX()'s will also dump the values of the compares and such. Doing this beats going in and changing all those AssertXXXs to include the compare-values in the message!

**Comment by PengTian0** [comment]
Can one of the admins verify this patch?

**Comment by johren** [comment]
@PengTian0: I'm not sure what specifically you are looking to verify?  Can you clarify?

File: test/deploy/rackhd_ha_resource_install.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+            # create mongo virtual ip resource
+            sb_net = self.get_southbound_network()
+            if sb_net:
+                ip = '{}.12{}'.format(sb_net, mongo)
+                vip_dict['mongo'].append(ip)
+                rsc_ip = 'mongo_addr_{}'.format(mongo)
+                self.assertTrue(self.configure_virtual_ip_resource(vmnum, ip, rsc_ip), "{} resource failure.".format(rsc_ip))
+            # colocate mongo and virtual IPs
+            mongo_cls = 'mongo{}'.format(mongo)
+            command = "crm configure colocation {} inf: {} {}".format(mongo_cls, rsc, rsc_ip)
```
Comment (larry-dean) [inline test/deploy/rackhd_ha_resource_install.py]: Not sure how hound is missing this but I'm pretty sure rsc_ip will be undefined if sb_net is None.

Comment (patelb10) [inline test/deploy/rackhd_ha_resource_install.py]: I can put those outside of 'if' condition

**Comment by johren** [comment]
Test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Comment by diontje** [comment]
test this please

**Comment by panpan0000** [comment]
test this please

**Comment by PengTian0** [comment]
ok to test

**Comment by changev** [comment]
ok to test

**Comment by changev** [comment]
test this please

**Comment by patelb10** [comment]
@larry-dean Can you verify now?

**Comment by panpan0000** [comment]
test this please

**Comment by panpan0000** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by PengTian0** [comment]
test this please

**Comment by johren** [comment]
Test this please

**Comment by johren** [comment]
test this please

**Comment by johren** [comment]
test this please

**Pull request merged**